### DISCUSSION TITLE
Why are breakpoints the building blocks of responsive design?

### DISCUSSION THREAD

**Discussion by jez9999** [twbs/bootstrap#37164] [Q&A]
Bootstrap's documentation states that breakpoints are the building blocks of responsive design.  However, I don't quite understand why this is so.  For example, I have a navbar that I want to 'shrinkwrap' to the inherent width of the content, and break into a burger menu only when the viewport gets too narrow to display the navbar at that width:

![navbar](https://user-images.githubusercontent.com/793830/190827375-c3ef5b8f-353a-4a7b-97db-6b74d74bbb61.png)

My understanding is that I can't really achieve this with Bootstrap, and that it's not designed to let me do it.  The navbar is supposed to collapse at a higher width that its 'shrinkwrapped' width, when there is still some black space to the right of it.  Why is this a better way to do responsive design?

To underscore what I'm talking about, look at the following navbar.  It's styled like this:

```
        <nav class="navbar fw-bold h5 rounded-5 navbar-expand d-inline-flex">
            <div class="container-fluid">
                <ul class="navbar-nav">
                    <li class="nav-item">
                        <NavLink class="nav-link" :to="{ name: 'Home' }">Home</NavLink>
                    </li>
                    [...]
                </ul>
            </div>
        </nav>
```

Because of `d-inline-flex`, it shrinkwraps so the background nearly fits the content, which is desired.  But Bootstrap either makes me break at a wider width than the navbar, like 992px with `.navbar-expand-lg`:

![navbar1](https://i.stack.imgur.com/xOdZl.gif)

... or not break at all, displaying a horizontal scrollbar instead of breaking just when the viewport gets too small:

![navbar2](https://i.stack.imgur.com/GeBkK.gif)

Neither of these are desirable to me.  Why doesn't Bootstrap cater for breaking to a burger menu at the natural content width?

**Comment by Edditoria** [comment]
There are 2 key points you may want to understand:

1. Use a wrapper `<div>` to style you menu content. (Inside `.container-fluid` in my example). Let Bootstrap does its job for `<navbar>` and `<container-fluid>`. You manage your own styles using wrappers.
2. Use Bootstrap's Toggler system for navbar. Please see [Responsive behaviors - Navbar - Bootstrap Docs](https://getbootstrap.com/docs/5.2/components/navbar/#responsive-behaviors). It contains a `<button>` as toggler and a `<div>` as a list of menu. Please make sure you use a good "id" to toggler only one element.

I created a demo for you: https://codepen.io/Edditoria/pen/NWMpmdJ

And, I added a `<div.navbar-toggler-wrapper>` for better style for the hamburger menu. It is not necessary but only "not going to hurt your eyes". 😸 

---

Bootstrap is highly rely on breakpoints for screen (media queries), rather than contents (element queries). And make sure it display properly on modern browsers, without or with-less Javascript.

For your case, you will need Javascript to get the width of the whole navbar. It can only be done **after** the browser renders the page (that users may use vary fonts, scales and other settings). I think Bootstrap does not have this feature.

Thanks.

**Reply by jez9999** [reply]
I already read that stuff, and your demo breaks on `navbar-expand-lg` which is significantly wider than the navbar content so it doesn't really address the question.

**Reply by Edditoria** [reply]
Oh sorry. I update and try to response for "Why is this a better way to do responsive design". Hope that other people can help to answer correctly.
